overlord/servicestate/quotas.go

```diff
@@ -0,0 +1,117 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package servicestate
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/snap/quota"
+)
+
+// AllQuotas returns all currently tracked quota groups in the state. They are
+// validated for consistency before returned and internally updated using
+// ResolveCrossReferences before returned.
+func AllQuotas(st *state.State) (map[string]*quota.Group, error) {
+	var quotas map[string]*quota.Group
+	if err := st.Get("quotas", &quotas); err != nil {
+		if err != state.ErrNoState {
+			return nil, err
+		}
+		// otherwise there are no quotas so just return nil
+		return nil, nil
+	}
+
+	// quota groups are not serialized with all the necessary tracking
+	// information in the objects, so we need to thread some things around
+	if err := quota.ResolveCrossReferences(quotas); err != nil {
+		return nil, err
+	}
+
+	// quotas has now been properly initialized with unexported cross-references
+	return quotas, nil
+}
+
+// GetQuota returns an individual quota group by name.
+func GetQuota(st *state.State, name string) (*quota.Group, error) {
+	allGrps, err := AllQuotas(st)
+	if err != nil {
+		return nil, err
+	}
+
+	// if the referenced group does not exist we return a nil group
+	return allGrps[name], nil
+}
+
+// UpdateQuotas will update the state quota group map with the provided quota
+// groups. The groups provided will be added on top of the current set of quota
```

Comment:
maybe for completeness: ... will be replace group states for the same name if present or be added on top of ...